feat(base-driver): generate declaration files

boneskull opened the pull request:
This was not as easy as the other two.  The main issue involves the `BaseDriver` class and is threefold:

1. `BaseDriver` and a valid external subclass thereof (a `Driver`) are not the same, because the former is essentially _abstract_. There is no notion of an abstract class in JS nor in TS for JS.
2. Splitting `BaseDriver` into several modules and "mixing in" some modules is just not well-understood by TS.
3. `AppiumDriver` inherits _most_ of `BaseDriver`, but overrides some functions with differing signatures. This is unsupported by TS.

To work around these problems, I had to:

1. Move the `Driver` interface into `@appium/types`, since other packages may depend on it.  We do not want circular dependencies, which is likely to occur otherwise.  Further, there is no way to tell TS for JS that a class _implements_ an interface or type.
2. Refactor the mixins (in `basedriver/commands/`) into "mixin classes", as understood by typescript. These are dynamic classes returned by functions which accept a class parameter, and _extend_ that class. The result is marginally better, but the implementation is ... well, see for yourself.  It is not possible to use declaration merging to just "trick" TS into thinking that `BaseDriver` has all of these extra methods; classes cannot be merged, and even if you could, you can't use declaration merging to "add" methods afaict.
3. `BaseDriver` is now just a class in a variable after the mixins have been applied. Mixins are applied to `BaseDriverCore`, and `BaseDriverCore` inherits from `DriverCore`. `AppiumDriver` _also_ inherits from `DriverCore`, since it does not override `BaseDriver` "properly".  `BaseDriverCore` itself is pretty small, containing only those functions which `AppiumDriver` defines as differing.

Otherwise:

- `BaseDriver` does not inherit from `Protocol`, which is just an empty class. `Protocol` was removed.
- All of the error classes are exported from `protocol/errors`, because TS cannot see them otherwise.
- Exported missing `GET_STATUS_COMMAND` from `protocol/index`
- Re-exported `normalizeBasePath` from `server` instead of `protocol`, since that's where it's defined
- Removed `index.d.ts`
-

boneskull committed:
feat(base-driver): generate declaration files

This was not as easy as the other two.  The main issue involves the `BaseDriver` class and is threefold:

1. `BaseDriver` and a valid external subclass thereof (a `Driver`) are not the same, because the former is essentially _abstract_. There is no notion of an abstract class in JS nor in TS for JS.
2. Splitting `BaseDriver` into several modules and "mixing in" some modules is just not well-understood by TS.
3. `AppiumDriver` inherits _most_ of `BaseDriver`, but overrides some functions with differing signatures. This is unsupported by TS.

To work around these problems, I had to:

1. Move the `Driver` interface into `@appium/types`, since other packages may depend on it.  We do not want circular dependencies, which is likely to occur otherwise.  Further, there is no way to tell TS for JS that a class _implements_ an interface or type.
2. Refactor the mixins (in `basedriver/commands/`) into "mixin classes", as understood by typescript. These are dynamic classes returned by functions which accept a class parameter, and _extend_ that class. The result is marginally better, but the implementation is ... well, see for yourself.  It is not possible to use declaration merging to just "trick" TS into thinking that `BaseDriver` has all of these extra methods; classes cannot be merged, and even if you could, you can't use declaration merging to "add" methods afaict.
3. `BaseDriver` is now an empty class which inherits from all mixins applied to `DriverCore`.  Due to syntactic restrictions, it must be a `class` statement in order to use the `@implements` tag, which causes TS to check that the result fully implements `Driver`.  `AppiumDriver` _also_ inherits from `DriverCore`, since it does not override `BaseDriver` "properly".  `BaseDriverCore` itself is pretty small, containing only those functions which `AppiumDriver` defines as differing.

Otherwise:

- `BaseDriver` does not inherit from `Protocol`, which is just an empty class. `Protocol` was removed.
- All of the error classes are exported from `protocol/errors`, because TS cannot see them otherwise.
- Exported missing `GET_STATUS_COMMAND` from `protocol/index`
- Re-exported `normalizeBasePath` from `server` instead of `protocol`, since that's where it's defined
- Removed `index.d.ts`
- Mixins use `@implements` as well. Note that `@implements` does not support `import('..')` syntax.

chore(base-driver): attempted to make types for mixins less awful

fixed a couple issues in the types in `driver.js`

core(base-driver): add capabilities types

- also move the bluebird config into the index
- fix references to `DriverOpts`
- add `assignServer()` to `BaseDriverCore`, used by `AppiumDriver`

chore(base-driver): fix more types

add missing reference to `@appium/types` in `tsconfig.json`

fix(base-driver): ensure getLogEvents is async

fix(base-driver): add some more type checks and fix logCustomEvent to be async

chore(base-driver): reformat

boneskull commented:
Current dependencies on/for this PR:
* 2.0
  * **PR #16594** <a href="https://app.graphite.dev/github/pr/appium/appium/16594" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #16595** <a href="https://app.graphite.dev/github/pr/appium/appium/16595" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #16596** <a href="https://app.graphite.dev/github/pr/appium/appium/16596" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  (this PR)
        * **PR #16597** <a href="https://app.graphite.dev/github/pr/appium/appium/16597" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

jlipps commented on packages/base-driver/lib/basedriver/commands/find.js:
```diff
@@ -1,51 +1,108 @@
-      this.log.debug(src);
+/**
+ *
+ * @param {ReturnType<import('./event').EventMixin>} Base
```
oh wow so we have to apply these in order?

boneskull replied:
Sort of.  It'd be one thing if these were _pure_ mixins, but they aren't.  They don't work standalone; they call out to other methods in `DriverCore` or maybe even other mixin classes.  You can't just apply them to any given class.  

It sucks, but I couldn't find another way to do this without making things so generic we don't know what's in the final class.  TS doesn't handle our use case very well.

If none of these actually call out to other mixins (I can't recall), then the order itself isn't important, but the types still are.  It's rather arbitrary right now.

boneskull replied:
another avenue I explored was just trying to trick TS into thinking that `BaseDriver` had a bunch of methods that were implemented elsewhere.  declaration merging (e.g., defining an `interface BaseDriver`) didn't work because TS generated broken declarations. module augmentation (`declare module './driver' { }`) didn't work because you cannot "add" stuff to a typed class.  so idk.  I was pulling my hair out.

all of this goes away, of course, if we just make a huge class.

boneskull replied:
also, nothing in `DriverCore` or `BaseDriverCore` is allowed to call out to a method defined in any of these mixins.  that--and the discrepancies introduced in `AppiumDriver` was why I had to move stuff around.

boneskull replied:
if typescript understood prototypes at all (it doesn't), I think this would be way easier.  but it's fundamentally based on classical OO and inheritance and all that garbage.  it makes some things difficult, and we found one of them.

```ts
class Foo {}

Foo.prototype = {bar: () => 'baz'};

let foo = new Foo();
foo.bar(); // TS238974239 error wtf no such method `bar`
```

jlipps commented on packages/base-driver/lib/basedriver/commands/find.js:
```diff
@@ -1,51 +1,108 @@
+/**
+ *
+ * @param {ReturnType<import('./event').EventMixin>} Base
+ * @returns {import('../driver').BaseDriverBase<import('@appium/types').TimeoutCommands & import('@appium/types').EventCommands & import('@appium/types').FindCommands>}
```
this is awful

jlipps replied:
can't we alias this type variable somewhere so it's not so ugly?

boneskull replied:
We can create a bunch of typedefs.  In TS this would be a bunch of `import` statements.

jlipps replied:
yeah, but imports are on their own up at the top, this is "in the middle" of code. it's ugly. typedefs are probably prettier.

jlipps commented on packages/base-driver/lib/basedriver/commands/find.js:
```diff
@@ -1,51 +1,108 @@
+     * @template {boolean} Mult
+     * @param {string} strategy
+     * @param {string} selector
+     * @param {Mult} mult
```
mult *should* just be a boolean (whether or not we are looking for multiple elements)

jlipps commented on packages/base-driver/lib/basedriver/commands/find.js:
```diff
@@ -1,51 +1,108 @@
+     * @param {string} selector
+     * @param {Mult} mult
+     * @param {string} [context]
+     * @returns {Promise<Mult extends true ? Element[] : Element>}
```
oh I see, yeah it's a tricky return type

boneskull replied:
Yeah.  I think I hate it `findElOrEls`, but too late now.

jlipps replied:
it's my fault, i guess i should have made driver authors implement both findEl and findEls instead /shrug

boneskull commented:
Failure is expected; #16597 should be the build to pass.

jlipps commented on packages/base-driver/lib/basedriver/commands/session.js:
```diff
@@ -1,153 +1,51 @@
 
-const commands = {};
-
-commands.createSession = async function createSession (w3cCapabilities1, w3cCapabilities2, w3cCapabilities) {
```
where did createSession, deleteSession, et al go?

jlipps commented on packages/base-driver/lib/basedriver/driver.js:
```diff
@@ -382,116 +115,267 @@ class BaseDriver extends Protocol {
+  async createSession (
+    w3cCapabilities1,
+    w3cCapabilities2,
+    w3cCapabilities,
```
is it a bit strange this one doesn't have a number? or that if it doesn't, that it's not first in the list?

jlipps commented on packages/base-driver/lib/basedriver/driver.js:
```diff
@@ -382,116 +115,267 @@ class BaseDriver extends Protocol {
-        return true;
+    this.originalCaps = _.cloneDeep(originalCaps);
+    this.log.debug(
+         `Creating session with W3C capabilities: ${JSON.stringify(
```
too much indentation

jlipps commented on packages/base-driver/lib/basedriver/driver.js:
```diff
@@ -382,116 +115,267 @@ class BaseDriver extends Protocol {
+     * @param {DriverData[]} [driverData]
+     * @returns {Promise<[string,object]>}
+     */
+  async createSession (
```
ah, here's createSession! i see where you put them now

jlipps commented on packages/base-driver/lib/helpers/capabilities.js:
```diff
@@ -20,6 +20,50 @@ function isW3cCaps (caps) {
   return false;
 }
 
+function fixCaps (originalCaps, desiredCapConstraints = {}, log) {
```
no type definitions in this function? is there a reason we're not ts-checking this file?

boneskull committed:
feat(appium): generate declaration files (#16597)

* feat(appium): generate declaration files

Configure TS to generate declaration files for `appium` package.

Most of these changes are just types getting added or changed via docstrings; in other places, typechecking has caught problems and I had to refactor code to address them.

- Removed `// @ts-check` pragma from all files in `lib/`, since now `checkJs: true` is in `tsconfig.json`.  It needs to stay in place elsewhere (I think) because only sources in `lib` have declarations generated.
- The `install`, `uninstall`, `update` and `run` functions in the base `ExtensionCommand` class were renamed and prefixed with a `_`; TS does not allow overridden functions with differing signatures.   It no longer exports a default.
- Removed `type` prop of options for `DriverCommand` and `PluginCommand` constructors, respectively (this value is derived from the `ExtensionConfig` `config` parameter instead)
- `AppiumDriver` now extends `DriverCore` instead of `BaseDriver`, since some of its function signatures differ from `BaseDriver`'s.
- Small refactor of `deleteSession` to make typechecker happy.
- `executeCommand` now calls to `BaseDriver` directly if `isUmbrellaCmd` is truthy.  This is one place where inheritance just kind of breaks, since `BaseDriver#executeCommand` returns something entirely different than `AppiumDriver#executeCommand` would otherwise.
- Removed `--delete-dir-on-start` from build command, since it may blast generated declarations.
- Reorganize declarations in `types/`; moved some of them out of `@typedef`s. `types.d.ts` becomes `index.d.ts`
- Note that this contains the declarations for `PluginClass` and `DriverClass`, which correspond to the types of extension "main" classes. This differs from the interfaces provided in `@appium/types` (e.g., `Driver`, `ExternalDriver`) in that these aren't just objects; they are classes with constructors and static members. I may end up moving some of this stuff into `@appium/types`, but I'm not sure how useful it will be.  Will see when I add types to `@appium/fake-driver`.

* chore(appium): remove transpile directives
this was ostensibly for gulp-babel

* chore(appium): remove redundant type declaration

* fix(appium): avoid potential unhandled rejections

- make types more readable

* chore(appium): fix capabilities types

- move `scripts/generate*` elsewhere
- add reference to `@appium/types` in `tsconfig.json`

* chore(appium): review updates for legibility
